src/interrupts/mod.rs

```diff
@@ -0,0 +1,577 @@
+    /// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+    ///     |interrupt_table| {
+    ///             let interrupt_handle = interrupt_table.register(Tim7, P1,
+    ///             || {
+    ///                 // Isr for interrupt `Tim7`
+    ///             }).expect("Interrupt already used");
+    ///
+    ///             /* Code that needs interrupt `Tim7` */
+    ///
+    ///             // Unregister interrupt and get back the ownership to `data`
+    ///             let data = interrupt_table.unregister(interrupt_handle);
+    ///             assert!(data.is_none());
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              mut isr: F)
+                              -> Result<InterruptHandle<()>, Error>
+        where F: FnMut() + 'a + Send
+    {
+        self.register_owned(irq, priority, (), move |_| {isr()})
+    }
+
+    /// Registers an interrupt with the lifetime of the `InterruptTable` and pass ownership of a variable `owned_data: T` that is passed
+    /// to the `isr(&mut T)` when the corresponding interupt `irq` occur.
+    /// 
+    /// The ownership of the data `owned_data` is returned in the `unregister()` function
+    ///
+    /// # Examples
+    /// ```
+    /// use stm32f7::interrupts::interrupt_request::InterruptRequest::Tim7;
+    /// use interrupts::Priority::P1;    
+    /// interrupts::scope(nvic, |irq| { hprintln!("Default handler: {}", irq) },
+    ///     |interrupt_table| {
+    ///             let data: SomeData = ...;
+    ///             let interrupt_handle = interrupt_table.register_owned(Tim7, P1, data,
+    ///             |owned_data| {
+    ///                 // Isr for interrupt `Tim7`
+    ///                 owned_data.do_anything();
+    ///             }).expect("Interrupt already used");
+    ///
+    ///             /* Code that needs interrupt `Tim7` */
+    ///
+    ///             // Unregister interrupt and get back the ownership to `data`
+    ///             let data = interrupt_table.unregister(interrupt_handle).unwrap();
+    /// });
+    /// ```
+    pub fn register_owned<F, T>(&mut self,
+                                 irq: InterruptRequest,
+                                 priority: Priority,
+                                 owned_data: T,
+                                 mut isr: F)
+                                 -> Result<InterruptHandle<T>, Error>
+        where   T: Send,
+                F: FnMut(&mut T) + 'a + Send
+    {
+        if unsafe{ISRS[irq as usize].is_some()} {
+            return Err(Error::InterruptAlreadyInUse(irq));
+        }
+        // Insert data only, when interrupt isn't used, therefore nobody reads the data => no dataraces
+        self.data[irq as usize] = unsafe {
+            transmute::<Option<*mut T>, Option<*mut ()>>(Some(Box::into_raw(Box::new(owned_data))))
```

Comment:
don't use transmute on raw pointers.

`Some(Box::into_raw(Box::new(owned_data)) as *mut ())` should totally suffice (or without the `Some` after the `Option` removal refactoring)